lib/OpenQA/Schema/Result/Jobs.pm

```diff
@@ -765,23 +766,34 @@ sub cluster_jobs {
     # fill dependency data; go up recursively if we have a directly chained or parallel parent
     my $parents = $self->parents;
   PARENT: while (my $pd = $parents->next) {
-        my $p = $pd->parent;
+        my $p         = $pd->parent;
+        my $parent_id = $p->id;
+        my $dep_type  = $pd->dependency;
 
-        if ($pd->dependency eq OpenQA::JobDependencies::Constants::CHAINED) {
-            push(@{$job->{chained_parents}}, $p->id);
+        if ($dep_type eq OpenQA::JobDependencies::Constants::CHAINED) {
+            push(@{$job->{chained_parents}}, $parent_id);
             # we don't duplicate up the chain, only down
             next;
         }
-        elsif ($pd->dependency eq OpenQA::JobDependencies::Constants::DIRECTLY_CHAINED) {
-            push(@{$job->{directly_chained_parents}}, $p->id);
+        elsif ($dep_type eq OpenQA::JobDependencies::Constants::DIRECTLY_CHAINED) {
+            push(@{$job->{directly_chained_parents}}, $parent_id);
             # duplicate also up the chain to ensure this job ran directly after its directly chained parent
             # notes: - We skip the children here to avoid considering "direct siblings".
             #        - Going up the chain is not required/wanted when cancelling.
-            $p->cluster_jobs(jobs => $jobs, skip_children => 1) unless $skip_parents || $args{cancelmode};
+            unless ($skip_parents || $args{cancelmode}) {
+                next if exists $jobs->{$parent_id};
+                die "Direct parent $parent_id needed to be cloned as well for the directly chained dependency "
```

Comment:
s/needed/needs/?

Reply:
This was supposed to conjunctive to imply that enforcing the action nevertheless is most likely not the ideal solution.

Reply:
But "needed" is simple past. Do you want `Direct parent $parent_id would need to be cloned as well for the directly chained dependency "`?

Reply:
For simplicity I've just changed it to "needs".